File: Practice/Problem Solving/Algorithms/Sock Merchant/SockMerchant.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.regex.*;

public class Solution {

    /**
     * Time Complexity: O(n).
     */

    static int sockMerchant(int n, int[] ar) {
        int pairs = 0;

        // Why HashSet?
        // Don't add duplicate values.
        // Don't care about order.
        Set<Integer> colors = new HashSet<>();

        for (int i = 0; i < n; i++) {
            // Time complexity of contain is O(1).
            if (!colors.contains(ar[i])) { 
                // Time complexity of add is O(1).
                colors.add(ar[i]);
            } else {
                pairs++;
                // Time complexity of remove is O(1).
                colors.remove(ar[i]);
            }
        }

        return pairs;
    }
}

File: Practice/Problem Solving/Algorithms/Sock Merchant/sock_merchant.cpp
#include <vector>
#include <unordered_set>

using namespace std;

/* Time complexity is O(n). */
int sockMerchant(int n, vector<int> ar)
{
    int pairs = 0;

    unordered_set<int> colors;

    // Add the first color.
    colors.insert(ar.at(0));

    for (int i = 1; i < n; i++)
    {
        // Time complexity of find is O(1).
        if (colors.find(ar.at(i)) == colors.end())
        {
            // Time complexity of insert is O(1).
            colors.insert(ar.at(i));
        }
        else
        {
            pairs++;

            // Time complexity of erase is O(1).
            colors.erase(ar.at(i));
        }
    }

    return pairs;
}